File: oxidize-golang/core/mesh/election_priority.go
package mesh

// Priority mirrors the deterministic Priority tuple used to rank nodes.
// Ordering: higher Clock wins; if equal, higher Seniority; if equal, higher
// CommandsSeen; if equal, lexicographically larger PeerID.
type Priority struct {
	Clock        uint64
	Seniority    uint64
	CommandsSeen uint64
	PeerID       string
}

// NewPriority constructs a priority tuple.
func NewPriority(clock, seniority, commandsSeen uint64, peerID string) Priority {
	return Priority{Clock: clock, Seniority: seniority, CommandsSeen: commandsSeen, PeerID: peerID}
}

// Less reports whether p ranks below other under the deterministic ordering.
func (p Priority) Less(other Priority) bool {
	if p.Clock != other.Clock {
		return p.Clock < other.Clock
	}
	if p.Seniority != other.Seniority {
		return p.Seniority < other.Seniority
	}
	if p.CommandsSeen != other.CommandsSeen {
		return p.CommandsSeen < other.CommandsSeen
	}
	return p.PeerID < other.PeerID
}

// electionDeclare is a recorded candidacy declaration for the active round.
type electionDeclare struct {
	priority Priority
}

// BullyElectionEngine mirrors the Rust BullyElection state machine with a
// deterministic Priority-based winner. It is additive alongside the simpler
// existing BullyElection type.
type BullyElectionEngine struct {
	LocalPeerID    string
	LocalSeniority uint64
	LocalCommands  uint64
	Clock          uint64
	electing       bool
	elected        bool
	master         string
	declares       map[string]Priority
	Completed      uint64
}

// NewBullyElectionEngine constructs an idle election engine.
func NewBullyElectionEngine(localPeerID string, localSeniority uint64) *BullyElectionEngine {
	return &BullyElectionEngine{
		LocalPeerID:    localPeerID,
		LocalSeniority: localSeniority,
		declares:       map[string]Priority{},
	}
}

// StartElection begins a new round with an incremented clock.
func (e *BullyElectionEngine) StartElection() {
	e.Clock++
	e.declares = map[string]Priority{}
	e.electing = true
	e.elected = false
}

// RecordDeclare records a remote candidacy if it belongs to the current round.
// Stale or future-clock declares are ignored.
func (e *BullyElectionEngine) RecordDeclare(clock, seniority, commandsSeen uint64, peerID string) {
	if !e.electing || clock != e.Clock {
		return
	}
	e.declares[peerID] = NewPriority(clock, seniority, commandsSeen, peerID)
}

// FinalizeElection computes the deterministic winner from all declares plus the
// local priority, transitions to elected, and returns the master peer ID.
func (e *BullyElectionEngine) FinalizeElection() (string, bool) {
	if !e.electing {
		return "", false
	}
	best := NewPriority(e.Clock, e.LocalSeniority, e.LocalCommands, e.LocalPeerID)
	for _, p := range e.declares {
		if best.Less(p) {
			best = p
		}
	}
	e.electing = false
	e.elected = true
	e.master = best.PeerID
	e.Completed++
	return best.PeerID, true
}

// IsMaster reports whether the local node is the elected master.
func (e *BullyElectionEngine) IsMaster() bool {
	return e.elected && e.master == e.LocalPeerID
}

// CurrentMaster returns the elected master peer ID, if any.
func (e *BullyElectionEngine) CurrentMaster() (string, bool) {
	if e.elected {
		return e.master, true
	}
	return "", false
}

// AcceptResult adopts a result broadcast if its clock is current or newer.
func (e *BullyElectionEngine) AcceptResult(clock uint64, master string) {
	if clock >= e.Clock {
		e.Clock = clock
		e.electing = false
		e.elected = true
		e.master = master
	}
}

// IncLocalCommands bumps the local commands-seen counter.
func (e *BullyElectionEngine) IncLocalCommands() { e.LocalCommands++ }

// RunElectionRound runs a full deterministic election using a capability graph:
// it starts the round, injects virtual declares for every peer in the graph,
// and finalizes. Mirrors run_election_round.
func RunElectionRound(e *BullyElectionEngine, graph *CapTopologyGraph) (string, bool) {
	e.StartElection()
	if graph != nil {
		graph.mu.Lock()
		for peerID, node := range graph.nodes {
			if peerID == e.LocalPeerID {
				continue
			}
			e.RecordDeclare(e.Clock, node.Seniority, node.CommandsSeen, peerID)
		}
		graph.mu.Unlock()
	}
	return e.FinalizeElection()
}

File: oxidize-python/oxidize_python/core/mesh/election_priority.py
"""Bully election mirroring oxidize-golang/core/mesh/election_priority.go.

Implements a deterministic ``Priority`` tuple ordering and a Bully election
state machine that selects a master peer.
"""

from __future__ import annotations

from dataclasses import dataclass

from .topology_extended import CapTopologyGraph


@dataclass(frozen=True)
class Priority:
    """Deterministic priority tuple used to rank nodes (mirrors Priority).

    Ordering: higher ``clock`` wins; if equal, higher ``seniority``; if equal,
    higher ``commands_seen``; if equal, lexicographically larger ``peer_id``.
    """

    clock: int = 0
    seniority: int = 0
    commands_seen: int = 0
    peer_id: str = ""

    def _key(self) -> tuple[int, int, int, str]:
        return (self.clock, self.seniority, self.commands_seen, self.peer_id)

    def less(self, other: "Priority") -> bool:
        """Report whether ``self`` ranks below ``other``."""
        return self._key() < other._key()


class BullyElectionEngine:
    """Bully election state machine with deterministic priority winner.

    Additive alongside the simpler existing ``BullyElection`` type.
    """

    def __init__(self, local_peer_id: str, local_seniority: int = 0) -> None:
        self.local_peer_id = local_peer_id
        self.local_seniority = local_seniority
        self.local_commands = 0
        self.clock = 0
        self._electing = False
        self._elected = False
        self._master = ""
        self._declares: dict[str, Priority] = {}
        self.completed = 0

    def start_election(self) -> None:
        """Begin a new round with an incremented clock."""
        self.clock += 1
        self._declares = {}
        self._electing = True
        self._elected = False

    def record_declare(
        self, clock: int, seniority: int, commands_seen: int, peer_id: str
    ) -> None:
        """Record a remote candidacy if it belongs to the current round."""
        if not self._electing or clock != self.clock:
            return
        self._declares[peer_id] = Priority(clock, seniority, commands_seen, peer_id)

    def finalize_election(self) -> tuple[str, bool]:
        """Compute the deterministic winner and transition to elected."""
        if not self._electing:
            return "", False
        best = Priority(
            self.clock, self.local_seniority, self.local_commands, self.local_peer_id
        )
        for p in self._declares.values():
            if best.less(p):
                best = p
        self._electing = False
        self._elected = True
        self._master = best.peer_id
        self.completed += 1
        return best.peer_id, True

    def is_master(self) -> bool:
        return self._elected and self._master == self.local_peer_id

    def current_master(self) -> tuple[str, bool]:
        if self._elected:
            return self._master, True
        return "", False

    def accept_result(self, clock: int, master: str) -> None:
        """Adopt a result broadcast if its clock is current or newer."""
        if clock >= self.clock:
            self.clock = clock
            self._electing = False
            self._elected = True
            self._master = master

    def inc_local_commands(self) -> None:
        self.local_commands += 1


def run_election_round(
    engine: BullyElectionEngine, graph: CapTopologyGraph | None
) -> tuple[str, bool]:
    """Run a full deterministic election using a capability graph.

    Starts the round, injects virtual declares for every peer in the graph and
    finalizes. Mirrors run_election_round.
    """
    engine.start_election()
    if graph is not None:
        for peer_id, node in graph._snapshot_nodes().items():  # noqa: SLF001
            if peer_id == engine.local_peer_id:
                continue
            engine.record_declare(
                engine.clock, node.seniority, node.commands_seen, peer_id
            )
    return engine.finalize_election()
